crypto/dilithium/p_dilithium3.c

```diff
@@ -0,0 +1,109 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```

Comment:
@[dkostic](https://github.com/dkostic) dkostic [27 days ago](https://github.com/awslabs/aws-lc/pull/637/files#r1002083002)
Regarding this:

> Dilithium offers both deterministic and randomized signing.

How do we handle this? In Dilithium's source code, the choice between deterministic and randomized signing is made based on a compile time flag which means we support only one of those two.

- Can we support only one signing method?
- If so, which one? Should we specify and document this?
- If we have to support both signing methods then we need to implement them both.
- Does the signing method implie the verifying method or is the verification always the same regadless of the signing method?

@[jakemas](https://github.com/jakemas) jakemas [16 days ago](https://github.com/awslabs/aws-lc/pull/637/files#r1010808091)
For now we will use deterministic signing.

Dilithium offers both deterministic and randomized modes of producing signatures. While the deterministic version may be better suited in constrained environments, or in devices that lack good sources of entropy, it can facilitate differential fault attacks that can be used to recover the private key. Randomized signing can mitigate against such attacks, but forces the security of the signature scheme to be dependent on a good source of randomness – as nonce reuse can lead to trivial secret key recovery.

While discussions are still underway about the best path forward, [a middle ground between the two methodologies known as hedging is being considered for use in Dilithium](https://groups.google.com/a/list.nist.gov/g/pqc-forum/c/1mQjngj_2Po/m/-p4RKXGQAwAJ), whereby the signature scheme produces randomized signatures that are seeded in part by the message and the key along side a source of randomness. We expect this method of hedged signing in all applications for which there is not a specific need for deterministic signing.

For interop we note that [liboqs](https://github.com/open-quantum-safe/liboqs/blob/55bc8cb38dbc01fadb324a95e00abb880cb03873/src/sig/dilithium/pqcrystals-dilithium_dilithium3_ref/config.h#L6) also uses deterministic signing.